[TwigBundle] Fix bug where namespaced paths don't take parent bundles in account

wesleylancel opened the pull request:
| Q | A |
| --- | --- |
| Branch? | 2.7 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #6919 |
| License | MIT |
| Doc PR |  |

Currently namespaced paths for templates such as `{% extends '@App/Layout/layout.html.twig' %}` do not work with bundles that have overruled templates using the `getParent()` method in another bundle. See attached ticket. This change prepends the path of the bundle implementing `getParent()` to the paths of the namespace of bundle returned as a parent.

jakzal commented:
We'll need a test case for this.

wesleylancel commented:
@jakzal I modified the TwigExtensionTest to include a test case for this situation.

wesleylancel commented:
AppVeyor ran on an outdated commit, might have to be restarted.

ro0NL commented on src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php:
```diff
@@ -92,6 +92,13 @@ public function load(array $configs, ContainerBuilder $container)
             if (is_dir($dir)) {
                 $this->addTwigPath($twigFilesystemLoaderDefinition, $dir, $bundle);
+
+                $classMock = new $class();
```
This is not really a mock :) what about `$object`?

wesleylancel replied:
I'll rename the variable. Don't think there are any options to retrieve the parent in this situation.

ro0NL replied:
Maybe you could call `registerBundles` once =/

wesleylancel replied:
I agree, and this won't work with a third level (ChildChildTwigBundle). How would you recommend calling `registerBundles` here? Is there any way to reach the kernel?

ro0NL replied:
`kernel` service is out of scope, so i guess not. Maybe you can kick in at `TwigBundle::createContainerExtension`, without breaking BC.

This is somewhat related to #18563 in terms of improving the ecosystem.

ro0NL replied:
Point is i think you need to traverse all possible parents, and register them for the current bundle `$dir`, so each parent works as expected.

jakzal replied:
I'm afraid we cannot instantiate the bundle class like this. Some bundles might have overridden the constructor (like the [JMSDebuggingBundle](https://github.com/schmittjoh/JMSDebuggingBundle/blob/master/JMSDebuggingBundle.php#L33)).

ro0NL replied:
What about this; as bundles are initialized before the container is ([ref](https://github.com/symfony/symfony/blob/973f29551255cac22b87e9e82ff93a280c9959c6/src/Symfony/Component/HttpKernel/Kernel.php#L114)) we can easily inject this kind of information. Currently `kernel.bundles` giving you only the FQCN is not flexible. We sometimes need the path, parent etc. So im thinking a special VO that replaces the FQCN is worth it... for now as additional parameter perhaps, but eventually to deprecate `kernel.bundles` being a list of FQCN's...

I would like this a lot!

wesleylancel replied:
@ro0NL Are you suggesting inject the entire `bundleMap` as a parameter after it has been initialized? This would also remove the need to initialize any classes. /cc @jakzal

ro0NL replied:
See #19594

wesleylancel replied:
Shouldn't the existing `bundleMap` property on the `Kernel` be able to provide the necessary data?

ro0NL replied:
Good point, havent noticed that one. Lets discuss further implementation details in #19594.

xabbuh replied:
What about using reflection and `newInstanceWithoutConstructor()` here?

chalasr replied:
@xabbuh some bundles make use of constructor arguments in their `getPath()` method. 
I tried to solve a similar issue in https://github.com/symfony/symfony/pull/18579

xabbuh replied:
That's true, but the only ways to solve that issue completely would be to either pass bundles to extension classes or to write some logic that determines the paths during cache warmup.

chalasr replied:
Couldn't a parameter returning the bundle map be envisaged?

xabbuh replied:
The issue is that you still would not have access to actual instances of the bundles (so this would not solve https://github.com/symfony/symfony/pull/19586#r94117685).

ro0NL commented on src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php:
```diff
@@ -92,6 +92,13 @@ public function load(array $configs, ContainerBuilder $container)
+                $classMock = new $class();
+                $parent = $classMock->getParent();
+
+                if ($parent) {
```
I guess `null !== $parent` is more verbose/consistent.

ro0NL replied:
That makes sense ;-) but do you agree on this approach? Im not sure how much we can rely on the default ordering of `kernel.bundles`, i guess we should correctly order them here for now.

wesleylancel replied:
Paths of child bundles will be prepended so they'll automatically come earlier than the original bundle's. Extra ordering should not be necessary in my opinion?

ro0NL replied:
If the order of `kernel.bundles` is correct (in a way that parents come before childs), then this is approach works. Basically i think we need a test that proves a more complex bundle hierarchy works as expected.

ro0NL replied:
The order is irrelevant, point is i think you need to traverse all possible parents, and register them for the current bundle `$dir`, so each parent works as expected.

wesleylancel replied:
Correct, I missed that. That seems impossible with the current code unless I can retrieve the complete bundleMap.

ro0NL replied:
Which you have... `$container->getParameter('kernel.bundles')` is a map of short- to classname.

wesleylancel replied:
The only way I could think of making this work is indeed, recursively instantiating the parent class, getting the parent bundle name, retrieve its class from the list of bundles, and so on. Is that what you would recommend?

ro0NL replied:
Yes, with performance in mind. Each bundle in the map should only be initialized once. Just keep in mind you're in a loop.

ro0NL replied:
Also if you have A>B>C, asking the parents for A should resolve the parents for B also. So if you ask for that later it's there.

ro0NL commented on src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php:
```diff
@@ -140,11 +147,23 @@ public function load(array $configs, ContainerBuilder $container)
+        $twigFilesystemLoaderDefinition->addMethodCall('prependPath', array($dir, $name));
+    }
+
+    private function stripBundleName($name)
```
What about `normalizeBundleName`?

ro0NL commented on src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/Fixtures/Bundle/TwigResourcesBundle/TwigResourcesBundle.php:
```diff
@@ -0,0 +1,13 @@
+
+use Symfony\Component\HttpKernel\Bundle\Bundle;
+
+class TwigResourcesBundle extends Bundle
```
Perhaps `ChildTwigBundle` is more verbose..

xabbuh commented:
@wesleylancel Do you think you can finish this one?

wesleylancel commented:
@xabbuh: I wish, as this is something that would be extremely useful for us. But, as pointed out by @ro0NL, bundles can extend infinitely A > B > C > D. My PR currently only checks for one level of extending. Ideally, I would be able to access the bundle map that is part of the kernel, but that's not available at this stage yet. If you could point me towards a possible solution I'd be happy to adjust my code accordingly.

ro0NL commented:
Seeing the related ticket this goes back to 2013. And quite some issues are involved. I've not experienced it myself, but the template problem seems real.

Either way allowing kernel introspection would be _extremely useful_ for everybody. Preferably with an immutable kernel. Practically we can move [this](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/Kernel.php#L493) a few lines up. Relying on a synthetic service before compilation.

We also can expose bundles thru parameters, we only need stringish values here. `BundleInterface` only defines name, path and namespace anyway. So passing a stringish bundle map could work. Any further details can be provided with `Bundle::build`.

But mayve @ogizanagi  has more idea's? As i also liked his approach with a contextual builder. Maybe it can be simplified.

ogizanagi commented:
Unfortunately #19646 was my only attempt to provide a naive feature for sharing some contextual information while the container is building, avoiding the need to rely on pseudo tmp synthetic services. Along with the `BootingKernel` and `BootingBundle` classes, it allowed to solves issues related to this one as a demonstration of the possibilities.
But despite the implementation is simple, it somehow looks overkilled for the few needs.

As you suggested, maybe exposing a simple representation of the bundle map and bundles informations in a kernel parameter could be the way. But I don't remember all the ins and outs and if it was easily feasible at this point. :/

ro0NL commented:
> But despite the implementation is simple, it somehow looks overkilled for the few needs.

Agree:+1: I feel the same about my own attempts now. Eventually having builder with context can be done in userland easily. But it doesnt solve it for SF.

If there arent any other insights, i'd say go for `%kernel.bundle_map%` :+1:

chalasr commented:
>  i'd say go for %kernel.bundle_map%

Not sure it's doable. Putting `Kernel::$bundleMap` in a new kernel parameter would just lead to

>  Unable to dump a service container if a parameter is an object or a resource.

ro0NL commented:
It's a stringish map, that's actually cached :) we're not passing bundle instances around here.

```
array(
  'name' => ['parent' => 'name2', 'namespace' => 'NS', 'path' => '/bundles/name' ],
  'name2' => //
)
```

Something like that i guess.

ro0NL commented:
Ie. looking at the issues, im not really sure we need true bundle instances here. Only metadata from bundle introspection. To build things properly.

chalasr commented:
See #21113

wesleylancel commented:
Nice @chalasr! I've subscribed to the PR and will update this one when it's merged.

xabbuh commented:
@wesleylancel *21113 has been merged.

wesleylancel commented:
@xabbuh Thanks for the heads up. Looking into rewriting this PR using the new parameter.

ogizanagi commented:
Is it still needed? @chalasr made the changes in the TwigExtension. What remains to do?

wesleylancel commented:
@ogizanagi I don't think that fixes registering extra paths for parent bundles.

chalasr commented:
>  I don't think that fixes registering extra paths for parent bundles.

Indeed, #21113 fixed the case for bundles overriding `getPath()`, handling bundle hierarchy still needs to be fixed

wesleylancel committed:
[TwigBundle] Fix bug where namespaced paths don't take parent bundles in account

wesleylancel commented:
@xabbuh @chalasr Just committed a first attempt using the new parameter. To the contrary of the earlier version this should take multiple levels of nested bundle into account. Feedback welcome.

chalasr commented:
@wesleylancel Looks good to me 👍  Not sure why but it seems you have a conflict

fabpot commented:
You should probably change the branch target to 2.7 instead of master.

wesleylancel commented:
Done.

xabbuh commented on src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php:
```diff
@@ -141,11 +145,41 @@ public function load(array $configs, ContainerBuilder $container)
+            $this->prependTwigPath($twigFilesystemLoaderDefinition, $dir, $name);
+        }
+
+        if (is_dir($dir = $parentBundle['path'].'/Resources/views')) {
```
should be `$bundleMetadata['path']`

wesleylancel replied:
@xabbuh: `$bundleMetadata` contains the entire array of bundles, `$parentBundle` is correct I think?

xabbuh replied:
Ah sorry, I meant `$bundleMetadata[$name]['path']`.

xabbuh commented on src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php:
```diff
@@ -141,11 +145,41 @@ public function load(array $configs, ContainerBuilder $container)
+        }
+
+        if (is_dir($dir = $parentBundle['path'].'/Resources/views')) {
+            $this->prependTwigPath($twigFilesystemLoaderDefinition, $dir, $name);
```
`$name` should be `$parentName`

wesleylancel replied:
@xabbuh The path should be registered with the bundle namespace that extends the parent bundle, so I think this should be allright too.

xabbuh replied:
But you want to register the path of the current bundle as a resource for the parent bundle's namespace.

wesleylancel replied:
Ah, yep, it's the wrong way around indeed. Will fix.

wesleylancel commented:
@xabbuh I think it should be the right way around now :)

chalasr commented:
@wesleylancel The calls of `realpath()` need to be removed [here](https://github.com/symfony/symfony/pull/19586/files#diff-00bc5ea153482178966f6617538a8e2eR282) and [here](https://github.com/symfony/symfony/pull/19586/files#diff-00bc5ea153482178966f6617538a8e2eR287) for the appveyor build to pass. Once fixed I would say we're good here

wesleylancel commented:
@chalasr Thanks, fixed.

dunglas commented:
Looks good to met.

👍

xabbuh commented:
👍 I checked this with the issues described in FriendsOfSymfony/FOSUserBundle#2376 and FriendsOfSymfony/FOSUserBundle#2378 and I can confirm that this patch fixes them.

Status: Reviewed

stof commented on src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php:
```diff
@@ -97,6 +97,10 @@ public function load(array $configs, ContainerBuilder $container)
             }
+
+            if (null !== $bundle['parent']) {
+                $this->prependParentPaths($twigFilesystemLoaderDefinition, $container, $name, $bundle['parent']);
```
Does this handle a chain of parents ?

stof replied:
Prepending paths looks wrong to me here, as it would overwrite user-configured paths for the same namespace too.
What you should do is collecting the child bundles for each bundle, and then adding paths for child bundles before adding the path of the bundle itself

wesleylancel replied:
@stof: Yes, this handles that. Should be covered with the changed test.

wesleylancel replied:
@stof: Did not now about that. That'll require some change to adding the normal paths too, I'll see how to best fix this. Is there an easy way to cover this in the test?

stof replied:
Add a path in the user-configured paths for the same namespace, and ensure that it ends up as the first path in the loader config for this namespace

ro0NL replied:
`prependParentPaths` was confusing at first. Maybe keep actual path registration explicit in `load()`

```php
while ($parent) {
   // add parent
}

// add self
```

Basically we dont need any path prepending or recursion, only adding paths with the right calling order.

kmelia commented:
@xabbuh great news! Could you tell us on which branches this patch will be applied ?

chalasr commented:
@kmelia Once merged, this will be available in the next patch releases of each maintained branch (i.e. from 2.7 to 3.2)  and the current development version (3.3).

kmelia commented:
@chalasr thanks!

stof commented on src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php:
```diff
@@ -203,8 +206,16 @@ public function testTwigLoaderPaths($format)
+            array(__DIR__.'/Fixtures/Bundle/ChildTwigBundle/Resources/views', 'Twig'),
+            array(__DIR__.'/Fixtures/Bundle/ChildChildTwigBundle/Resources/views', 'ChildTwig'),
+            array(__DIR__.'/Fixtures/Bundle/ChildChildTwigBundle/Resources/views', 'Twig'),
+        ), $prependPaths);
```
The way these tests are written does not ensure that paths are added in the right order for each namespace. there is nothing guaranteeing that the combinations of appended paths and prepended paths is actually giving us the right priority.

What we should do to have reliable tests is rebuilding the internal structure of ``Twig_Loader_Filesystem::$path`` (i.e. the result of the configuration) and ensuring that it looks good for each namespace.

wesleylancel replied:
@stof: I think whatever gets prepended last will have the most priority. If the array above is in this order, the correct path will end up on top. However, this doesn't keep those user-configured paths in mind. I'm not sure where these get added to the loader, but if they get prepended after the bundle paths get prepended, it should work all right? Not sure if that's testable in the existing test though.

If the Twig internals of that loader do need to be changed, I don't think this can be merged before that is resolved?

wesleylancel replied:
@stof: What if we move these lines https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php#L84-L90 to here https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php#L109 and change the method call to `prependPath` instead of `addPath`? This should make sure that user-defined paths from config take precedence over anything else.

stof replied:
the issue is that the way the test is written does not allow to quickly validate what the final order will be when looking at it. This is why I think the test should not be written this way. We don't want to assert 2 separate lists, as we care about a single list (the one being used at runtime).

thus, the test is even confusing when looking at it, because the list of ``prependPath`` here is in reverse order compared to the order of paths in the loader (as prepending adds at the beginning while you append in ``$prependPaths``), making it even harder to validate/understand the assertion.

wesleylancel replied:
@stof I'll see if I can collect all the paths before actually adding them to the loader so they can all be tested together in order.

wesleylancel committed:
Build all paths first, then add to loader

wesleylancel commented:
@stof I've added a separate commit so it's easier to see what I've changed (can be squashed later). A hierarchy of bundles is now built first, after that the paths are added to the loader. I'm not sure if I could simplify the code, especially the building of the hierarchy.

I could improve the test by asserting the array per namespace if that helps for readability?

xabbuh commented:
I think an easier implementation could be to have two kids. The first would collect path information for bundles we override without actually registering them. The second would then first register the additional paths from child bundles followed by paths the bundle provides itself.

wesleylancel commented:
@xabbuh: Isn't that what I'm doing now? First the hierarchy is built and then the paths are registered.

stof commented on src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php:
```diff
@@ -139,13 +144,69 @@ public function load(array $configs, ContainerBuilder $container)
+                );
+            }
+
+            if (is_dir($dir = $container->getParameter('kernel.root_dir').'/Resources/'.$name.'/views')) {
```
Note to maintainers: when merging to newer branches, we need to take care of the FileExistenceResource objects here (git conflicts should remind us about this btw)

stof commented:
👍

dunglas commented:
👍

fabpot commented:
Thank you @wesleylancel.

homersimpsons commented:
@fabpot Do you have any hint about the new release (containing this fix) date?

chalasr commented:
@homersimpsons today :) see https://github.com/symfony/symfony/pull/21258
